homeassistant/components/mazda/button.py

```diff
@@ -0,0 +1,153 @@
+from dataclasses import dataclass
+from typing import Any
+
+from pymazda import Client as MazdaAPIClient
+
+from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from . import MazdaEntity
+from .const import DATA_CLIENT, DATA_COORDINATOR, DOMAIN
+
+
+async def handle_button_press(
+    client: MazdaAPIClient,
+    key: str,
+    vehicle_id: int,
+    coordinator: DataUpdateCoordinator,
+) -> None:
+    """Handle a press for a Mazda button entity."""
+    api_method = getattr(client, key)
+
+    try:
+        await api_method(vehicle_id)
+    except Exception as ex:
+        raise HomeAssistantError(ex) from ex
+
+
+async def handle_refresh_vehicle_status(
+    client: MazdaAPIClient,
+    key: str,
+    vehicle_id: int,
+    coordinator: DataUpdateCoordinator,
+) -> None:
+    """Handle a request to refresh the vehicle status."""
+    await handle_button_press(client, key, vehicle_id, coordinator)
+
+    await coordinator.async_refresh()
+
+
+@dataclass
+class MazdaButtonRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    # Suffix to be appended to the vehicle name to obtain the button name
+    name_suffix: str
+
+
+@dataclass
+class MazdaButtonEntityDescription(
+    ButtonEntityDescription, MazdaButtonRequiredKeysMixin
+):
+    """Describes a Mazda button entity."""
+
+    # Function to determine whether the vehicle supports this button, given the coordinator data
+    is_supported: Callable[[dict[str, Any]], bool] = lambda data: True
+
+    async_press: Callable[
+        [MazdaAPIClient, str, int, DataUpdateCoordinator], Awaitable
+    ] = handle_button_press
+
+
+BUTTON_ENTITIES = [
+    MazdaButtonEntityDescription(
+        key="start_engine",
+        name_suffix="Start Engine",
+        icon="mdi:engine",
+    ),
+    MazdaButtonEntityDescription(
+        key="stop_engine",
+        name_suffix="Stop Engine",
+        icon="mdi:engine-off",
+    ),
+    MazdaButtonEntityDescription(
+        key="turn_on_hazard_lights",
+        name_suffix="Turn On Hazard Lights",
+        icon="mdi:hazard-lights",
+    ),
+    MazdaButtonEntityDescription(
+        key="turn_off_hazard_lights",
+        name_suffix="Turn Off Hazard Lights",
+        icon="mdi:hazard-lights",
+    ),
+    MazdaButtonEntityDescription(
+        key="start_charging",
+        name_suffix="Start Charging",
+        icon="mdi:ev-station",
+        is_supported=lambda data: data["isElectric"],
+    ),
+    MazdaButtonEntityDescription(
+        key="stop_charging",
+        name_suffix="Stop Charging",
+        icon="mdi:ev-station",
+        is_supported=lambda data: data["isElectric"],
+    ),
```

Comment:
Does the api know the state of the engine/hazard lights/charging?

Reply:
Charging should be a switch in that case

Reply:
Please update the opening text of the PR to reflect the changes.

Thanks 👍

Reply:
Updated.